File: example/tasks/web/src/refusals.ts
// The half of the client that is worth more than the calls that work: the ones
// that do not compile.
//
// Every `@ts-expect-error` below is an assertion. If the generator ever widens
// a type so that one of these becomes legal, tsc fails on the unused
// suppression — so this file is a test, and it fails in the direction that
// matters (a guard that stops guarding).
//
// Each case is something the server would answer 400 to. The point of
// generating from the schema rather than from the OpenAPI document is that they
// are answered here instead, before the request is sent.

import { listTasks, getComment, type Transport } from './api/client.gen';
import { taskMutations } from './api/queries.gen';
import { useMutation } from '@tanstack/react-query';

export function refusals(request: Transport) {
  // @ts-expect-error — "titel" is not a column of tasks.
  void listTasks(request, { where: { titel: 'typo' } });

  // @ts-expect-error — position is sortable and selectable, but not filterable...
  void listTasks(request, { where: { position: 3 } });
  // ...and the capabilities it did declare still work:
  void listTasks(request, { sort: 'position', select: ['position'] });

  // @ts-expect-error — "reviewing" is not one of the status enum's values.
  void listTasks(request, { where: { status: 'reviewing' } });

  // @ts-expect-error — contains is a pattern operator and comment_count is a number.
  void listTasks(request, { where: { comment_count: { contains: '3' } } });

  // @ts-expect-error — list_id is not nullable, so there is no null to test for.
  void listTasks(request, { where: { list_id: { isnull: true } } });

  // @ts-expect-error — description is searchable but not sortable.
  void listTasks(request, { sort: 'description' });

  // labels is an array column. What it accepts is containment...
  void listTasks(request, { where: { labels: { has: 'urgent' } } });
  void listTasks(request, { where: { labels: { hasany: ['urgent', 'blocked'] } } });
  void listTasks(request, { where: { labels: ['a', 'b'] } });

  // @ts-expect-error — ...and `contains` is the text operator, not this one.
  void listTasks(request, { where: { labels: { contains: 'urgent' } } });

  // @ts-expect-error — `has` takes one element, not the array.
  void listTasks(request, { where: { labels: { has: ['urgent'] } } });

  // @ts-expect-error — array ordering is not offered, so there is no gt.
  void listTasks(request, { where: { labels: { gt: ['a'] } } });

  // @ts-expect-error — labels declared Filterable, deliberately not Sortable.
  void listTasks(request, { sort: 'labels' });

  // @ts-expect-error — a scalar column has no containment operator.
  void listTasks(request, { where: { title: { has: 'x' } } });

  // @ts-expect-error — password_hash is hidden: it has no spelling at all here.
  void listTasks(request, { select: ['password_hash'] });

  // @ts-expect-error — comments declare no expandable relation.
  void getComment(request, 'id', { expand: ['task'] });

  // @ts-expect-error — the item endpoint has no ?select, and rejects one.
  void getComment(request, 'id', { select: ['body'] });
}

/**
 * The mutation layer inherits the same refusals, because its variables are the
 * body types rather than a second spelling of them.
 */
export function writeRefusals(request: Transport) {
  const tasks = taskMutations(request);
  const create = useMutation(tasks.create);
  const update = useMutation(tasks.update);
  const complete = useMutation(tasks.complete);

  // @ts-expect-error — workspace_id is owned by a hook, so the create body has
  // no such property and there is nothing for the server to ignore.
  create.mutate({ list_id: 'l', title: 't', description: '', workspace_id: 'w' });

  // @ts-expect-error — comment_count is read-only: it is filterable and
  // sortable, and still has no spelling in a patch body.
  update.mutate({ id: 'a', body: { comment_count: 3 } });

  // @ts-expect-error — an item verb's variables carry the id, not just the body.
  complete.mutate({ note: 'done' });

  // And the calls that do work, so the refusals above are not passing because
  // the whole shape is wrong:
  create.mutate({ list_id: 'l', title: 't', description: '' });
  update.mutate({ id: 'a', body: { title: 'renamed' } });
  complete.mutate({ id: 'a', body: { note: 'done' } });
}

export async function narrowing(request: Transport) {
  const page = await listTasks(request, { select: ['title'] });
  const first = page.items[0];
  if (first === undefined) return;

  first.id; // the primary key is always returned
  first.title;
  // @ts-expect-error — status was not selected, so it is not on the row.
  first.status;

  const expanded = await listTasks(request, { expand: ['list'], per_page: 1 });
  const task = expanded.items[0];
  if (task === undefined) return;
  task.list.name; // present, not optional, because it was expanded

  const plain = await listTasks(request, { per_page: 1 });
  const unexpanded = plain.items[0];
  if (unexpanded === undefined) return;
  // @ts-expect-error — nothing was expanded, so there is no list to read.
  unexpanded.list.name;
}

File: example/tasks/mobile/lib/refusals.dart
// What does *not* compile, asserted.
//
// Each case below is a request the server would answer 400 to, written out and
// suppressed with `// ignore:`. The `unnecessary_ignore` lint is on in
// `analysis_options.yaml`, so if the generator ever widened one of these types
// the analyser would report the suppression as unnecessary and the gate would
// fail. The guard fails when it stops guarding — which is the property the
// TypeScript client gets from `@ts-expect-error`, in the language that has no
// such annotation.
//
// Nothing here is meant to be called. It exists to be analysed.
//
// ignore_for_file: unused_element

library;

import 'api/client.gen.dart';

// A column that never declared .Filterable() has no spelling in `where`.
// `tasks.author_id` is a reference the schema chose not to make filterable.
void _unfilterableColumn() {
  // ignore: undefined_named_parameter
  const TaskWhere(authorId: Cond(eq: 'u1'));
}

// Pattern operators need a text column. `comment_count` is an int, so `Cond`
// has no `contains`, and the server would refuse it.
void _patternOnANumber() {
  // ignore: undefined_named_parameter
  const TaskWhere(commentCount: Cond(contains: '3'));
}

// A null test on a column that cannot be null is not a question SQL answers.
// `list_id` is required, so its condition type has no `isNull`.
void _nullTestOnARequiredColumn() {
  // ignore: undefined_named_parameter
  const TaskWhere(listId: Cond(isNull: true));
}

// An enum column compares against its own value set, not against any string.
void _bareStringAgainstAnEnum() {
  // ignore: const_constructor_param_type_mismatch, argument_type_not_assignable
  const TaskWhere(status: Cond(eq: 'todo'));
}

// And not against another column's enum either.
void _theWrongEnum() {
  // ignore: const_constructor_param_type_mismatch, argument_type_not_assignable
  const TaskWhere(status: Cond(eq: TaskPriority.high));
}

// A value type the column does not have.
void _theWrongValueType() {
  // ignore: const_constructor_param_type_mismatch, argument_type_not_assignable
  const TaskWhere(commentCount: Cond(eq: 'three'));
}

// A hidden column has no spelling anywhere: not in the row, not in `select`,
// not as a filter. `users.password_hash` never leaves the process.
void _hiddenColumn(User user) {
  // ignore: undefined_getter
  user.passwordHash;
}

// A column that did not declare .Sortable() is not in the sort vocabulary.
// `tasks.list_id` is filterable and not sortable.
void _unsortableColumn() {
  // ignore: undefined_enum_constant
  TaskListParams(sort: [TaskSort.listId.asc]);
}

// `select` names this resource's columns. Another resource's are a different
// enum, so they do not typecheck.
void _anotherResourcesColumn() {
  // ignore: list_element_type_not_assignable
  const TaskListParams(select: [CommentColumn.body]);
}

// `expand` offers the relations the schema marked expandable, in the direction
// they are served. A task has no `comments` expansion.
void _anExpansionTheResourceDoesNotOffer() {
  // ignore: undefined_enum_constant
  TaskListParams(expand: [TaskExpand.comments]);
}

// The item endpoint declares no `select`: `rest` registers it with
// RejectUnknownQueryParameters, so offering one would generate requests the
// server refuses.
void _selectOnTheItemEndpoint() {
  // ignore: undefined_named_parameter
  const TaskGetParams(select: [TaskColumn.title]);
}

// Read-only columns are the database's or a hook's. `comment_count` is
// maintained by a trigger, so a create body has no property for it.
void _writingAReadOnlyColumn() {
  // ignore: undefined_named_parameter
  const TaskCreate(listId: 'l1', title: 't', description: '', commentCount: 0);
}

// A create body cannot omit a column the database has no default for.
void _omittingARequiredColumn() {
  // ignore: missing_required_argument
  const TaskCreate(listId: 'l1');
}

// An immutable column is settable once, at create. `workspace_id` is
// ReadOnly on every table here, so a patch has no method to write it.
void _patchingAnImmutableColumn() {
  // ignore: undefined_method
  TaskPatch().workspaceId('w2');
}

// Passing null to a column that is not nullable writes a NULL the database
// would reject, so the patch method does not accept one.
void _clearingARequiredColumn() {
  // ignore: argument_type_not_assignable
  TaskPatch().title(null);
}

// A patch writes columns, not relations: the expansion is a response shape.
void _patchingARelation() {
  // ignore: undefined_method
  TaskPatch().list('l2');
}
